### ISSUE TITLE
Using external memory allocator with PyTorch

### ISSUE THREAD

**Issue by jakirkham** [pytorch/pytorch#43144]
## 🚀 Feature

It would be useful to configure PyTorch to use an external memory allocator for its allocations.

## Motivation

When working on GPUs, memory can be a somewhat limited resources. Particularly when using multiple libraries each handling their own memory. In this case it is possible for libraries to "compete" for memory and exhaust it more quickly as a result. In these use cases, it's helpful to have a single memory allocator that can be shared across libraries. This reduces some of the friction of memory allocations between libraries.

## Pitch

It would be great to have some mechanism for users of PyTorch to specify an external memory allocator possibly like [CuPy's]( https://docs.cupy.dev/en/stable/reference/memory.html#changing-memory-pool ), [Numba's]( https://numba.readthedocs.io/en/stable/cuda/external-memory.html ), or similar.

## Alternatives

Feel free to suggest 😉 

## Additional context

We frequently receive reports from users of RAPIDS that they can run out of memory when using RAPIDS (say for preprocessing) and pass data to PyTorch for DL. In this context having a more unified memory management story should help users transition smoothly between libraries and reuse memory already freed (though still with the pool) of RAPIDS Memory Manager ([RMM]( https://github.com/rapidsai/rmm )).

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel

**Comment by zasdfgbnm** [comment]
We got the same issue when someone was trying to integrate libtorch with [Amber](https://ambermd.org/)

**Comment by jakirkham** [comment]
(that would be a really cool use case to enable too and timely I might add :)

**Comment by harrism** [comment]
[XGBoost recently did this as well.](https://github.com/dmlc/xgboost/pull/5873) The fact that they already had an external memory manager interface made it much easier to use RMM with it (same was true of Numba).

**Comment by EvenOldridge** [comment]
Would also allow true copyless data transfer between RAPIDS and PyTorch via dlpack if they share memory spaces.  (my understanding is that there's a copy if the space isn't shared)

**Comment by harrism** [comment]
I in no way intend to water down the value of a PyTorch external memory interface, but I just want to provide the facts here:

> Would also allow true copyless data transfer between RAPIDS and PyTorch via dlpack if they share memory spaces. (my understanding is that there's a copy if the space isn't shared)

This is only true in the PyTorch Tensor --> dlpack -> cuDF case and only for column-major tensors. Going the other way requires copies unless the DF only has a single column, because multicolumn DFs have a separate allocation per column.

**Comment by slayton58** [comment]
It's been an issue with DGL and their PyT backend too, as well as with DALI.

Another alternative would be to offer an external API for torch's allocators for use in external codes.

**Comment by colesbury** [comment]
An external allocator is possible, but there are some complications:

- It has to avoid synchronization in the common alloc/dealloc case or PyTorch perf will suffer a lot
- Multiprocessing requires getting the pointer to the underlying allocation for sharing memory across processes. That either has to be part of the allocator interface, or you have to give up on sharing tensors allocated externally across processes.

Exposing the PyTorch allocator is also possible.

Maybe @ngimel has thoughts on this.

**Comment by soumith** [comment]
I'm generally supportive of the direction, but we do need to figure out some details.

What @colesbury said would be technical needs.
How it might manifest is in the form of import side-effects.

```
import foo    # here, foo imports and sets torch to an external allocator that is suboptimal
import torch

out = mymodel(x) # this line runs much slower if `foo` is imported because of the suboptimal overriden allocator
```

Detecting a suboptimal allocator I think isn't technically possible (other than profiling).
So, when an external allocator is enabled, a reasonable thing for pytorch to do is to enable warnings to the user that something suboptimal might happen.

My concern is around whether we have an external memory allocator that we all agree is reasonably performant, that we allow overrides to (or allow overrides to without emitting a warning).

Any thoughts?

**Comment by harrism** [comment]
I would suggest that exposing PyTorch's allocator for external use and allowing specifying an external allocator for use in PyTorch are two very different things. This issue is requesting the latter.

In my opinion compatibility in the ecosystem will benefit more if all libraries in the ecosystem provide a way to specify an external allocator, because this enables savvy users to use the same allocator across their entire workflow within the ecosystem. If many parts of the ecosystem provide this, but one library only provides a way to use it's allocator externally, then that potentially forces whole workflows to use that allocator.

While performance is often a core motivator, there are other reasons to choose an allocator. The main motivation in this case is to avoid OOM errors when different libraries in your workflow each reserve large pools of memory. Other use cases range from development reasons (debugging features like logging and memory leak detection and profiling features like heap profilers) to systems reasons (custom allocators that support specific hardware, distributed task managers, etc.). Also, it's common to layer different allocation strategies to accommodate or tune for specific allocation patterns of an application or workflow. 

If libraries in the ecosystem all provide a way to specify external allocators, then I'm not sure it matters if all parts of the ecosystem agree on what the best allocator is, since a given workflow or use case can select the allocator or allocators that satisfy their needs, and each library can provide sensible defaults. That said, if many can agree on a common allocator that certainly helps.

I am biased to RMM (linked above) since I maintain it and I think it provides a great set of features and excellent performance, but to reiterate the point of this issue, if PyTorch exposes an external memory interface, any external allocator can be used. 

Happy to expound on the benefits of RMM though if useful. :smile:

**Comment by jrhemstad** [comment]
> My concern is around whether we have an external memory allocator that we all agree is reasonably performant, that we allow overrides to (or allow overrides to without emitting a warning).

When a library exposes an interface for user-defined allocation, it's typically understood to be a _caveat emptor_ situation. If a user picks a bad custom allocator then that is their prerogative and PyTorch should have no obligation to protect the user from themselves. 

Granted, exposing a pluggable allocator interface is a potential footgun. However, users providing a custom allocator typically know what they are doing. The risk of someone shooting themselves in the foot is outweighed by the fact that a pluggable allocator interface enables usecases and solves real problems that users are facing today.

**Comment by soumith** [comment]
@jrhemstad the catch22 is not when a user picks a bad custom allocator. It is when a library that the user is using opaquely picks a bad custom allocator. In that situation, the user might not even be aware that this is happening under the hood, they only see bad performance and attribute it ot "something something pytorch bad..."

**Comment by kkraus14** [comment]
> @jrhemstad the catch22 is not when a user picks a bad custom allocator. It is when a library that the user is using opaquely picks a bad custom allocator. In that situation, the user might not even be aware that this is happening under the hood, they only see bad performance and attribute it to "something something pytorch bad..."

I would argue that we're already in a catch22 here as right now we can end up with multiple different caching/pool allocators and they all end up starving for memory but PyTorch throws the OOM error and a user still attributes it to "something something pytorch bad...".

In practice, CuPy has exposed a memory allocator hook and has their own default Memory Pool implementation. Maybe someone from the CuPy team can share any insights they've gained from supporting this?

**Comment by jakirkham** [comment]
cc @emcastillo @kmaehashi @leofang (from CuPy to see if they have thoughts here)
cc @gmarkall @seibert (from Numba in case they have thoughts as well)

**Comment by gmarkall** [comment]
Some general experience from Numba: exposing a hook for using an external memory allocator has worked well. The hook was first proposed in [a Numba Enhancement Proposal (NBEP)](https://numba.readthedocs.io/en/stable/proposals/external-memory-management.html) and received some feedback from the community before being finalised. The hook allows for setting the memory manager at startup, then using that manager for all Numba-related allocations and deallocations throughout the duration of execution. The main documentation for Numba's external management, which goes into some detail about considerations when using it, is [here](https://numba.readthedocs.io/en/stable/cuda/external-memory.html).

External memory managers can be used by implementing a plugin that is passed to the hook - the interface specifies methods for initialization, allocation of device memory, memory pinning, etc. that Numba can call when it needs to allocate - plugins are implemented by subclassing the [BaseCUDAMemoryManager class](https://numba.readthedocs.io/en/stable/cuda/external-memory.html#numba.cuda.BaseCUDAMemoryManager). RMM has [an implementation of this interface](https://github.com/rapidsai/rmm/blob/branch-0.15/python/rmm/rmm.py#L161) so a Numba / RMM user doesn't need to write any code (beyond setting the Numba memory manager to the RMM implementation using the hook) to use RMM for Numba.

Since it was implemented, Numba uses the RMM allocator whenever cuDF is used (because cuDF sets Numba's allocator when it is imported) - there haven't been any major issues resulting from this, as far as I'm aware. More generally, there haven't been any significant issues or regressions come up with the use of an external allocator so far.

RMM provides better performance for than Numba's internal allocation strategy, so when users have issues with memory management performance I generally direct them to use RMM via the hook, which usually improves things. This also takes some pressure off the idea of working to improve the performance of Numba's memory management strategy, which is a nice benefit from a maintenance/workload perspective.

**Comment by datametrician** [comment]
@soumith what do you think are next steps here? 
Enabling this would really help our cybersecurity use-cases where our GPUs are pushing memory limits with large batches of log files. 
@BartleyR would you share more, please?

> I would argue that we're already in a catch22 here as right now we can end up with multiple different caching/pool allocators and they all end up starving for memory but PyTorch throws the OOM error and a user still attributes it to "something something pytorch bad...".

@kkraus14 alluded to it already as well.

**Comment by BartleyR** [comment]
We have a pipeline for parsing raw logs where we read into a cuDF series via custreamz (but it generalizes to any streaming pipeline with GPU support). We do some light processing of the raw logs then tokenize (using cuDF) before we move to PyTorch tensors. Inference is done using a transformer model, and tensors are returned to cuDF for post-processing (e.g., removing split tokens, reforming split logs). This presents an issue because we need both the RAPIDS Memory Manager (RMM) and the PyTorch allocator right now. When we're done with inference, we don't really need those tensors until the next stream comes in. We're doing a lot of manual management right now (removing tensors, etc.). In the ideal case though, a unified/single memory manager would let us utilize every spare piece of GPU memory much more efficiently. 

The short version though is what @kkraus14 said. This is the case we're ended up with, and the OOM error is being thrown from PyTorch. In this case, we know what the root issue is (it's not necessarily PyTorch, it's the competing allocators).

**Comment by leofang** [comment]
> In practice, CuPy has exposed a memory allocator hook and has their own default Memory Pool implementation. Maybe someone from the CuPy team can share any insights they've gained from supporting this?

Sorry for joining the party late. Since the PFN folks are busy at preparing the CuPy release, let me say a few words on what I understand for the memory allocator issue. 

We would love to see all major Python GPU libraries (CuPy, Numba, PyTorch, cuDF, etc) freely accept each other's memory allocator for exactly the same reason as @kkraus14 mentioned. I think as @jrhemstad pointed out the advanced users typically want a unified memory pool for all Python applications so that we can eliminate competitions among multiple allocators --- If one constantly runs into OOM, worrying about the allocation/deallocation efficiency for a particular choice of allocator becomes meaningless IMHO.

For this, @emcastillo from PFN has made quite some effort: the PR #33860 (already merged into PyTorch's main branch a while ago) allows CuPy (via cupy/cupy#3126) and other libs to use PyTorch's memory pool. I think @jakirkham's proposal here is to allow the opposite direction: let PyTorch use any lib's memory pool. I think it would be a nice use case for PyTorch to be able to use say RMM (which CuPy and Numba both support). I am happy to see this happening.

**Comment by emcastillo** [comment]
I think we can collaborate in isolating the memory allocator and make a pluggable interface, @asi1024 got quite familiar with the c10 code so we can definitely work together in this.

**Comment by soumith** [comment]
@datametrician i'd let @csarofeen @ptrblck and @ngimel figure out next steps from the PyTorch side. I believe they are figuring out the subtle details and exact plan

**Comment by EvenOldridge** [comment]
This would be very helpful for NVTabular's accelerated dataloaders as well.  Having to explicitly define the memory sizes for cuDF beforehand limits what we can do in terms of taking advantage of the memory available after the model is loaded.  What we have works, but this would be more flexible.

**Comment by ezyang** [comment]
A good external allocator story would also help rationalize some of our internal diffs like https://github.com/pytorch/pytorch/pull/42006 and https://github.com/pytorch/pytorch/pull/43951

cc @kimishpatel @dreiss

**Comment by quasiben** [comment]
It's been awhile since this has been updated and I'm curious if there is still motivation from PyTorch devs to work on this.  Additionally, are there things RAPIDS folks can help with here ?

**Comment by mcarilli** [comment]
I just started work on adding cudaMallocAsync to Pytorch: https://github.com/pytorch/pytorch/pull/65365

https://github.com/pytorch/pytorch/pull/65365 won't support external allocators at first, but in the future, the interface where Pytorch chooses between its native allocator and cudaMallocAsync could also allow choosing functions from some user-supplied, external shared library (that implements the same interface). Pytorch's allocator interface isn't set in stone either, in principle I could change it to match some external "standard" if that standard is durable, common, and supports all Pytorch's use cases.

**Comment by jrhemstad** [comment]
>  I could change it to match some external "standard" if that standard is durable, common, and supports all Pytorch's use cases.

How about just using the stream-ordered interface/semantics of `cudaMallocAsync/cudaFreeAsync`? The only caveat I would push for is to require the size of the allocation be passed to the `deallocate` function. (Though even this isn't required at the cost of runtime overhead of storing a `map<void*,size_t>` to map an allocation to its size.) 

Otherwise, the stream-ordered allocation interface of `cudaMallocAsync` is the exact same interface used by [RMM](https://github.com/rapidsai/rmm#device_memory_resource) and soon to be superseded by [`cuda::memory_resource`](https://github.com/NVIDIA/libcudacxx/pull/158) in libcu++.

**Comment by harrism** [comment]
I agree it would be best for the ecosystem if the EAI matches `cuda::memory_resource` (@jrhemstad do you think alignment should be included?). I second the request to require the size of the allocation in the API of the deallocate function. But most important is for the stream-ordering semantics of allocation and deallocation to match cudaMallocAsync, otherwise interoperability will be difficult or require excess synchronization.

**Comment by mcarilli** [comment]
also, do you have any recommendations on how the external allocator can be "pluggable"? Having the pytorch allocator's interface functions fish symbols from a user-supplied .so was my first idea, but I don't know the taxonomy of how other libraries have done this.

(for example, afaict https://github.com/cupy/cupy/pull/4537 exposes cudaMallocAsync alloc/free directly in Python, but that's a non-starter for Pytorch, we need the implementation to be chosen behind the allocator's interface in C++.)

**Comment by jrhemstad** [comment]
> How does that benefit implementations enough to be part of a universal interface?

Well, it's not really a "universal" interface. It's a C interface and follows C-like patterns. If you want a "universal" C++ interface, look to the STL [`Allocator`](https://en.cppreference.com/w/cpp/named_req/Allocator) concept or  [`std::pmr::memory_resource`](https://en.cppreference.com/w/cpp/memory/memory_resource), both of which require passing the size of allocation at deallocation. 

> And if it's super beneficial to implementations, why wasn't cudaFreeAsync designed that way?

Because it's a C interface and `free` doesn't take the deallocation size either. In most `malloc` implementations, they store the size of the allocation in some padding bytes prior to the returned `ptr`. You obviously can't do that with device memory, so you need to maintain some kind of `map<void*,size_t>` like I mentioned. It's not a lot of overhead to do that, but it is _some_ and largely pointless overhead in C++. 

In C++, memory should always be allocated as part of some container/RAII class that already stores the associated allocation size. Therefore, providing the size of the allocation at deallocation is free and eliminates the need for the `map<void*,size_t>`. 

> do you have any recommendations on how the external allocator can be "pluggable"?
> we need the implementation to be chosen behind the allocator's interface in C++

Sounds like you might want to consider building on [`cuda::memory_resource`](https://github.com/NVIDIA/libcudacxx/pull/158). It's a polymorphic interface (`allocate/deallocator` are virtual) inspired by [`std::pmr::memory_resource`](https://en.cppreference.com/w/cpp/memory/memory_resource) for plugging in allocators at runtime.

**Comment by zhuzilin** [comment]
Hi there, I'm also interested in building an custom allocator for pytorch. Before having the plugin interfaces, I wonder if using `LD_PRELOAD` for `cudaMalloc` and `cudaFree` is a good walk around. I've tried to make an example with cnmem in [zhuzilin/pytorch-malloc](https://github.com/zhuzilin/pytorch-malloc). Could you share some suggestions with this idea?

**Comment by jakirkham** [comment]
@zhuzilin, that's neat! Thanks for sharing 😄

Have you tried with [RMM]( https://github.com/rapidsai/rmm )? 🙂

**Comment by harrism** [comment]
While this may work, it's a bit hacky and probably [a security risk that wouldn't be allowed in some production environments](https://security.stackexchange.com/questions/63599/is-there-any-way-to-block-ld-preload-and-ld-library-path-on-linux).

**Comment by zhuzilin** [comment]
@jakirkham I was using cnmem because it only has one header file and it's much easier to do a PoC :P

**Comment by koparasy** [comment]
It's been a while since this has been updated and I'm curious if there is still motivation from PyTorch devs to work on this?
